endpoints/openrtb2/auction.go

```diff
@@ -300,6 +299,183 @@ func validateBanner(banner *openrtb.Banner, impIndex int) error {
 	return nil
 }
 
+// fillAndValidateNative validates the request, and assigns the Asset IDs as recommended by the Native v1.2 spec.
+func fillAndValidateNative(n *openrtb.Native, impIndex int) error {
+	if n == nil {
+		return nil
+	}
+
+	var nativePayload nativeRequests.Request
+	if err := json.Unmarshal(json.RawMessage(n.Request), &nativePayload); err != nil {
+		return err
+	}
+
+	if err := validateNativeContext(nativePayload.Context, impIndex); err != nil {
+		return err
+	}
+	if err := validateNativePlacementType(nativePayload.PlcmtType, impIndex); err != nil {
+		return err
+	}
+	if err := fillAndValidateNativeAssets(nativePayload.Assets, impIndex); err != nil {
+		return err
+	}
+
+	// TODO #218: Validate eventtrackers once mxmcherry/openrtb has been updated to support Native v1.2
+
+	serialized, err := json.Marshal(nativePayload)
+	if err != nil {
+		return err
+	}
+	encoded, err := json.Marshal(string(serialized))
+	if err != nil {
+		return err
+	}
+	n.Request = string(encoded)
+	return nil
+}
+
+func validateNativeContext(c nativeRequests.ContextType, impIndex int) error {
+	if c < 1 || c > 3 {
+		return fmt.Errorf("request.imp[%d].native.request.context must be in the range [1, 3]. Got %d", impIndex, c)
+	}
+	return nil
+}
+
+func validateNativePlacementType(pt nativeRequests.PlacementType, impIndex int) error {
+	if pt < 1 || pt > 4 {
+		return fmt.Errorf("request.imp[%d].native.request.plcmttype must be in the range [1, 4]. Got %d", impIndex, pt)
+	}
+	return nil
+}
+
+func fillAndValidateNativeAssets(assets []nativeRequests.Asset, impIndex int) error {
+	if len(assets) < 1 {
+		return fmt.Errorf("request.imp[%d].native.request.assets must be an array containing at least one object.", impIndex)
+	}
+
+	for i := 0; i < len(assets); i++ {
+		// Per the OpenRTB spec docs, this is a "unique asset ID, assigned by exchange. Typically a counter for the array"
+		// To avoid conflict with the Request, we'll return a 400 if the Request _did_ define this ID,
+		// and then populate it as the spec suggests.
+		if err := validateNativeAsset(assets[i], impIndex, i); err != nil {
+			return err
+		}
+		assets[i].ID = int64(i)
```

Comment:
From the spec: "Typically a counter for the array." This enforces that it will be a counter. I don't know of any use cases for ID being anything else, so its probably good. But potentially will break something that stretches the spec here.

Reply:
yeah... to be honest I'm not 100% sure how to handle this. I saw the following:

> Unique asset ID, assigned by exchange.

Since Prebid is the exchange, in this context, I don't _think_ we're breaking the spec here. We just decided that the "typical" implementation is the one we'll be using.

I decided to return 4xx on IDs sent by the Publisher just to prevent all sorts of nasty logic/confusion. As a general rule, we never _overwrite_ anything which the Publisher gives us. So if we allowed IDs, then we'd have to make sure that the ones we added weren't duplicates, which gets... messy.

Decided to just reject those requests to keep things simple.

Reply:
I am fine with that then.